src/_pytest/capture.py

```diff
@@ -77,19 +81,29 @@ def __init__(self, method):
         self._method = method
         self._global_capturing = None
         self._current_item = None
+        self._atexit_funcs = []  # type: List[Callable]
+        atexit.register(self._atexit_run)
```

Comment:
I think the tmpfile management can be isolated in a separate class, which would greatly simplify the code:

```python
class Streams(IntEnum):

    STDIN = 0
    STDOUT = 1
    STDERR = 2


@attr.s
class TemporaryFilePool:

    _tmpfiles = attr.ib(default=attr.factory(dict))

    def obtain_tmpfile(self, stream_id):
        try:
            tmpfile = self._tmpfiles[stream_id]            
        except KeyError:
            if stream_id == Streams.STDIN:
                tmpfile = open(os.devnull, "r")
            else:                
                f = TemporaryFile()
                with f:
                    tmpfile = safe_text_dupfile(f, mode="wb+")
            self._tmpfiles[stream_id] = tmpfile
        
        assert not tmpfile.closed
        return tmpfile

    def close(self):
        for f in self._tmpfiles.values():
            f.close()
        self._tmpfiles.clear()
```

We should then make `CaptureManager` create and maintain the `tmpfile_pool`, and pass it around to `FDCaptureBinary`. `CaptureManager` should then register the `atexit` handle itself, so these 3 lines would become:

```python
self._tmpfile_pool = TemporaryFilePool()
atexit.register(self._tmpfile_pool)
```

This design also allows us to change the pool implementation, so we could even possibly control it via a config variable and get back the old behavior (I'm not actually suggesting it, just mentioning that this is then possible).

Reply:
Oh I also realize now that we don't respect `basetemp` when creating `TemporaryFiles`...